diff -up nfs4-acl-tools-0.3.3/libnfs4acl/nfs4_ace_from_string.c.save nfs4-acl-tools-0.3.3/libnfs4acl/nfs4_ace_from_string.c
--- nfs4-acl-tools-0.3.3/libnfs4acl/nfs4_ace_from_string.c.save	2012-01-31 14:24:49.738876000 -0500
+++ nfs4-acl-tools-0.3.3/libnfs4acl/nfs4_ace_from_string.c	2012-01-31 14:25:07.074388000 -0500
@@ -86,9 +86,10 @@ free_fields(char *fields[NUMFIELDS])
 {
 	int i;
 
-	for (i = 0; i < NUMFIELDS; i++)
-		if (fields[i] != NULL)
-			free(fields[i]);
+	for (i = 0; i < NUMFIELDS; i++) {
+		free(fields[i]);
+		fields[i] = NULL;
+	}
 }
 
 int
@@ -100,14 +101,14 @@ parse_alloc_fields(char *buf, char *fiel
 	if (!buf)
 		return -EINVAL;
 
-	memset(fields, 0, sizeof(fields));
+	memset(fields, 0, sizeof(char *) * NUMFIELDS);
 
 	for (i = 0; buf[i] != '\0'; i++) {
 		if (buf[i] == ':')
 			count++;
 	}
 	if (count != 3)
-		goto out_free;
+		return -EINVAL;
 
 	for (i = 0; i < NUMFIELDS; i++) {
 		field = strsep(&buf, ":");
@@ -150,7 +151,7 @@ struct nfs4_ace * nfs4_ace_from_string(c
 	/* parse_alloc_fields had split up ace_buf so now we copy it to bufp */
 	bufp = malloc(strlen(ace_buf) + 1);
 	if (!bufp)
-		goto out_free;
+		goto out;
 	strcpy(bufp,ace_buf);
 
 	ret = parse_alloc_fields(bufp, fields);
